File: command_recognition.py

class RecognizedCommand:
    '''Objects of this class will be returned to you from the 'recognize'
    function of the 'CommandRecognizer' class. Stores information about the command name, 
    its attributes, and the part of the message where it was received'''
    def __init__(self, commandName = None, fullSentence = None, commandAttributes = None):
        self.commandName = commandName #str
        self.commandAttributes = commandAttributes #str
        self.fullSentence = fullSentence #str


class CommandData:
    '''Objects of this class are used to store information about a particular command.'''
    def __init__(self, commandName, shouldHaveAttributes, keywords):
        self.commandName = commandName #str
        self.shouldHaveAttributes = shouldHaveAttributes #bool
        self.keywords = keywords #[str1, str2, ...]


class CommandsListHandler:
    '''Class exists to manage list that contains commands information
    and list that stores conjuctions and words that should be deleted'''
    #HOW COMMANDS LIST WILL LOOK LIKE
    ALL_COMMANDS = {
        #"hello": CommandData("hello", False, ["hello", "hi"]),
        #"music": CommandData("music", True, ["music"]),
        #"name_of_cmd": CommandData("name_of_cmd", bool(should_have_attrs), ["keyphrase1", "keyphrase2"])
    }

    TRASH_WORDS = {
        "conjunctions": [],  #["and", "after that", "also"], # helps to understand if new command starts
    }


    @staticmethod
    def create_new_command(commandName, shouldHaveAttributes, keywords):
        '''Creates new command for this exactly session.
        By the first attribute you should give your command a name (string),
        in the future command recognixer will return exactly this name.
        By the second attribute you should give boolean value, true if your command 
        should have attributes after its occurrence in the message.
        Third parameter is list of keywords, that you think command must starts with.
        Examples: "hello", False, ["hello", "hi"];
                  "music", True, ["turn on music", "i want to listen to music", "music"]
        from sentence "hello i want to listen to music despacito by luis" 
        you'll get recognized commands
        "hello" with no attributes 
        "music" with attribute "despacito by luis"
        '''
        CommandsListHandler.ALL_COMMANDS[commandName] = CommandData(commandName, shouldHaveAttributes, keywords)

    @staticmethod
    def add_conjuction(conjuction):
        '''Add word/phrase that will be used as potential separator
        for commands. For example, "music despacito hello" will return 
        command name "music" with attribute "despacito hello", but if 
        before this you add conjusction "and also" with this function,
        message "music despacito and also hello" will return 
        command "music" with attribute "despacito" and command "hello"
        ''' 
        CommandsListHandler.TRASH_WORDS["conjunctions"].append(conjuction)

    @staticmethod
    def get_object(data_name, key = None):
        '''This function used by exclusively developer of this module. Just ignore'''
        if data_name == "ALL_COMMANDS": data = CommandsListHandler.ALL_COMMANDS
        elif data_name == "TRASH_WORDS": data = CommandsListHandler.TRASH_WORDS
        else: raise NameError("Object named '{}' wasn't found".format(data_name)) 
        if key == None:
            return data
        if key in data:
            return data[key]
        else:
            raise KeyError("Key '{}' wasn't found in '{}'.".format(key, data_name))


class CommandRecognizer:
    @staticmethod
    def recognize(message):
        recognizedCommands = []
        possibleCommands = {
            "shouldHaveAttributes": {},
            "notHaveAttributes": {}
        }

        ALL_COMMANDS = CommandsListHandler.get_object("ALL_COMMANDS")

        for command in ALL_COMMANDS:
            for keyword in ALL_COMMANDS[command].keywords:
                haveAttributes = "shouldHaveAttributes" if ALL_COMMANDS[command].shouldHaveAttributes \
                            else "notHaveAttributes"
                if keyword in message:
                    try:
                        possibleCommands[haveAttributes][command] = [*possibleCommands[haveAttributes][command], keyword]
                    except:
                        possibleCommands[haveAttributes][command] = [keyword]
        # possibleCommands["notHaveAttributes"] ~= {"command_name": ["keyword_1", "keyword_2", ...], ...}
        
        message = message.split()
        word_count = len(message)
        detected = 0

        for index, word in enumerate(message):
            if CommandRecognizer.new_command_starting(possibleCommands["shouldHaveAttributes"], message, index, leading=(not detected))["state"]:
                message[index] = "|{}".format(message[index])
                detected += 1
            
        message = ' '.join(message)

        #now commands that should have attributes, have '|' in front of them
        message = message.split('|')
        first_cmd_index = len(message[0].split())

        # now lets remove conjuctions from the end of command, 
        # as its probably not the part of command attributes
        
        TRASH_WORDS = CommandsListHandler.get_object("TRASH_WORDS")

        for index, item in enumerate(message):
            was_deleted = True 
            while was_deleted: # keep deleting while we can
                was_deleted = False
                for trash in TRASH_WORDS["conjunctions"]:
                    if ' '.join(item.split()[-len(trash.split()):]) == trash:
                        item = ' '.join(item.split()[:-len(trash.split())])
                        message[index] = item
                        was_deleted = True
                        
                    if item == '':
                        was_deleted = False
                        break

        # because of conjuctions at the end was deleted, now commands, that should
        # have attributes, divised from cmds, that shouldn't, with conjuctions in front of cmds 
        # that shouldn't have attributes (like "|<cmd_keyphrase> <attributes> <conjuctions>
        # <cmd_that_shouldn't_have_attr> |<cmd_keyphrase> <attributes> <etc.> ...")
        message = ('|'.join(message)).split('|')
        message = ' |'.join(message).split()

        detected = 0

        for index, word in enumerate(message):

            leading = ((not detected and index < first_cmd_index) or first_cmd_index == word_count)
            
            command_starts = CommandRecognizer.new_command_starting(possibleCommands["notHaveAttributes"], 
                message, index, leading = leading)

            if command_starts["state"]:
                parsed = command_starts["keyphrase"].split()
                parsed[0] = '|'+parsed[0]
                parsed[len(parsed)-1] = parsed[len(parsed)-1]+'|'
                message[index:index+len(parsed)] = parsed
                detected = 1

        message = ' '.join(message).split('|')

        #deleting conjuctions in front of cmds that dont have attrs
        for index, item in enumerate(message):
            was_deleted = True 
            while was_deleted: # keep deleting while we can
                was_deleted = False
                for trash in TRASH_WORDS["conjunctions"]:
                    if ' '.join(item.split()[-len(trash.split()):]) == trash:
                        if trash == 'и':
                            print(message)
                        item = ' '.join(item.split()[:-len(trash.split())])
                        message[index] = item
                        was_deleted = True
                        if trash == 'и':
                            print(message)
                    if item == '':
                        was_deleted = False
                        break

        for item in message:
            command = CommandRecognizer.command_recognized(item.strip(), possibleCommands)
            if command:
                recognizedCommands.append(RecognizedCommand(commandName = command["name"], 
                    fullSentence = item, commandAttributes = 
                    command["attributes"].strip() if ALL_COMMANDS[command["name"]].shouldHaveAttributes else None))

        return recognizedCommands

    @staticmethod
    def command_recognized(sentence, possibleCommands):
        '''Function to check if sentence starts with any command keyword/keyphrase'''
        for attrMode in possibleCommands:
            for possibleCommand in possibleCommands[attrMode]:
                for keyword in possibleCommands[attrMode][possibleCommand]:
                    if sentence.startswith(keyword):
                        return {
                            "name": possibleCommand,
                            "attributes": sentence.replace(keyword, "", 1)
                        }

        return None

    @staticmethod
    def new_command_starting(possibleCommands, sentence, startIndex, leading = False):
        '''Function to check if new command is starting'''
        for command in possibleCommands:
            for keyphrase in possibleCommands[command]:
                if ' '.join(sentence[startIndex:]).startswith(keyphrase) and \
                (startIndex == 0 or leading or CommandRecognizer.have_conjuctions_in_front(sentence, startIndex)):
                    return {"state": True, "keyphrase": keyphrase}

        return {"state": False, "keyphrase": None}

    @staticmethod
    def have_conjuctions_in_front(sentence, index):
        '''Function to check if word have conjuctions in front'''
        for conj in CommandsListHandler.get_object("TRASH_WORDS", "conjunctions"):
            if ' '.join(sentence[index-len(conj.split()):index]) == conj:
                return True

        return False

File: command_recognition.dart
class CommandData {
  final String commandName;
  final bool shouldHaveAttributes;
  final List<String> keywords;
  
  CommandData(this.commandName, this.shouldHaveAttributes, this.keywords);
}

class RecognizedCommand {
  final String commandName;
  final String commandAttributes;
  final String fullSentence;
  
  RecognizedCommand(this.commandName, this.commandAttributes, this.fullSentence);
}

class CommandsListHandler {
  static Map<String, CommandData> allCommands = {
    
  };
  
  static Map<String, Set<String>> trashWords = {
    "conjuctions": {},
    "to_remove": {},
  };
  
  static void create_new_command(String commandName, bool shouldHaveAttrs, List<String> keywords) {
    CommandsListHandler.allCommands[commandName] = CommandData(commandName, shouldHaveAttrs, keywords);
  }
  
  static void add_conjuction(String conjuction) {
    CommandsListHandler.trashWords["conjuctions"].add(conjuction);
  }
  
  static void add_word_to_delete(String word) {
    CommandsListHandler.trashWords["to_remove"].add(word);
  }
}

class CommandRecognizer {
  static List<RecognizedCommand> recognize(String message) {
    
    Map<String, Map<String, List<String>>> _possibleCommands = {
      "shouldHaveAttributes": {},
      "notHaveAttributes": {},
    };
    
    Map<String, CommandData> _allCommands = CommandsListHandler.allCommands;
    
    for(String command in _allCommands.keys) {
      String attributeMode = _allCommands[command].shouldHaveAttributes ? "shouldHaveAttributes" : "notHaveAttributes";
      for(String keyword in _allCommands[command].keywords) 
        if (message.contains(keyword)){
          if (_possibleCommands[attributeMode].containsKey(command))
          _possibleCommands[attributeMode][command] = [
              ..._possibleCommands[attributeMode][command],
              keyword
            ];
          else 
            _possibleCommands[attributeMode][command] = [keyword]; 
        }
    }
    
    List _messageStruct = message.split(" ");
    int _wordCount = _messageStruct.length;
    bool _detected = false;
    int _index = 0;
    
    for (String word in _messageStruct){
      if (CommandRecognizer.new_command_starting(
        _possibleCommands["shouldHaveAttributes"], _messageStruct, 
        _index, leading: (!_detected))["state"]) {
                _messageStruct[_index] = "|${_messageStruct[_index]}";
                _detected = true;
      }
      _index ++;
    }
    
    
    message = _messageStruct.join(" ");
    _messageStruct = message.split("|");
    int _firstCmdIndex = _messageStruct[0].split(" ").length;
    
    Map<String, Set> _trashWords = CommandsListHandler.trashWords;
    
    // just deleting conjuctions in front of already recognized commands
    _index = -1;
    for(String item in _messageStruct){
      bool wasDeleted = true;
      int itemLen = item.split(" ").length;
      
      while (wasDeleted) { // keep deleting while we can (for the case when 2+ conj-s at the end)
        wasDeleted = false;
        _index ++;
        for(String trash in _trashWords["conjuctions"]){
          int trashLen = trash.split(" ").length;
          
          if(trashLen >= itemLen)
            continue;
          
          if (item.split(" ").sublist(itemLen-trashLen, itemLen).join(" ") == trash){
            item = item.split(" ").sublist(0, itemLen-trashLen).join(" ");
            _messageStruct[_index] = item;
            wasDeleted = true;
            itemLen -= trashLen;
          }
          if (item == ""){
            wasDeleted = false;
            break;
          }
        }
        
      }
    } // end of deleting
    
    
    _messageStruct = _messageStruct.join("|").split("|").join(" |").split(" "); // doing stuff u know...
    _detected = false;
    
    // recognizing commands that shoudnt have attrs
    _index = 0;
    for(String word in _messageStruct){
      bool leading = (!_detected && (_index < _firstCmdIndex) || _firstCmdIndex == _wordCount);
      Map cmdStarts = CommandRecognizer.new_command_starting(
        _possibleCommands["notHaveAttributes"], _messageStruct, _index, leading: leading
        );
      
      if(cmdStarts["state"]){
        List parsed = cmdStarts["keyphrase"].split(" ");
        int parsedLen = parsed.length;
        
        parsed[0] = "|" + parsed[0];                     // placing borders 
        parsed[parsedLen-1] = parsed[parsedLen-1] + "|"; // of command

        for(int i = _index; i < _index+parsedLen; i ++)
          _messageStruct[i] = parsed[i-_index];

        _detected = true;
      }

      _index ++;
    }
    

    _messageStruct = _messageStruct.join(" ").split("|");
    
    for(int i = 0; i < _messageStruct.length; i++)  // deleting spaces from
      _messageStruct[i] = _messageStruct[i].trim(); // the end of items
    
    // deleting conjuctions in front of commands that shouldnt have attrs
    _index = -1;
    for(String item in _messageStruct){
      bool wasDeleted = true;
      int itemLen = item.split(" ").length;
      _index ++;
      while (wasDeleted) { // keep deleting while we can (for the case when 2+ conj-s at the end)
        wasDeleted = false;
        for(String trash in _trashWords["conjuctions"]){
          int trashLen = trash.split(" ").length;
          
          if(trashLen >= itemLen)
            continue;
          
          if (item.split(" ").sublist(itemLen-trashLen, itemLen).join(" ") == trash){
            item = item.split(" ").sublist(0, itemLen-trashLen).join(" ");
            _messageStruct[_index] = item;
            wasDeleted = true;
            itemLen -= trashLen;
          }
          if (item == ""){
            wasDeleted = false;
            break;
          }
        }
        
      }
    } // end of deleting
    
    List<RecognizedCommand> _recognizedCommands = [];
    
    
    for(String item in _messageStruct){
      
      Map command = CommandRecognizer.command_recognized(item.trim(), _possibleCommands);
      if (command.containsKey("name")){
        _recognizedCommands.add(RecognizedCommand(command["name"], command["attributes"], item));
      }
    }

    return _recognizedCommands;
  }

  static Map command_recognized(String sentence, Map possibleCommands) {
    for(String attrMode in possibleCommands.keys)
      for(String possibleCommand in possibleCommands[attrMode].keys)
        for(String keyword in possibleCommands[attrMode][possibleCommand])
          if (sentence.startsWith(keyword)) {
            return {
              "name": possibleCommand,
              "attributes": sentence.replaceFirst(keyword, "").trim()
            };
          }

    return {};
  }
  
  static Map new_command_starting(Map possibleCommands, List<String> sentence, int index, {leading = false}) {
    for(String command in possibleCommands.keys){
      for(String keyphrase in possibleCommands[command]){
        if (sentence.sublist(index).join(" ").startsWith(keyphrase) && (index == 0 || 
            leading || CommandRecognizer.have_conjuctions_in_front(sentence, index)))
          return {"state": true, "keyphrase": keyphrase};
      }
    }
      
    return {"state": false, "keyphrase": null};
  }

  static bool have_conjuctions_in_front(List<String> sentence, int index) {
    for(String conj in CommandsListHandler.trashWords["conjuctions"])
      if (sentence.sublist(index-conj.split(" ").length, index).join(" ") == conj)
        return true;
    return false;
  }
}

void main() {
  List cmds = [
    ["привет", false, ["привет", "хай"]],
    ["музыка", true, ["включи музыку", "музыка"]],
    ["заметка", true, ["новая заметка", "заметка", "добавь новую заметку"]],
    ["погода", false, ["какая сейчас погода", "погода"]]
  ];
  
  for(List cmd in cmds)
    CommandsListHandler.create_new_command(cmd[0], cmd[1], cmd[2]);

  for (String item in ["и", "потом", "затем", "после этого", "а ещё"])
      CommandsListHandler.add_conjuction(item);
  
  for(var a in CommandRecognizer.recognize("привет включи музыку деспасито и добавь новую заметку погода хороша и музыка из шрека")){
    print("${a.commandName} ${a.commandAttributes}");
  }
}
